File: src/api/feriados.ts
import { callCommand } from './client'
import type { CivilDate } from './types'

/** See `docs/11-contratos-tauri.md` §5.13. */

export type OrigenFeriado = 'Api' | 'Manual'

export interface Feriado {
  fecha: CivilDate
  nombre: string
  tipo: string | null
  /** A sync never overwrites a `Manual` row: what the user loaded by hand wins. */
  origen: OrigenFeriado
}

export interface FeriadoInput {
  fecha: CivilDate
  nombre: string
}

export interface FeriadoSyncResult {
  agregados: number
  total: number
  /** Years the provider could not be reached for; the calendar stays as it was. */
  aniosConError: number
}

export function listFeriados(anio: number): Promise<Feriado[]> {
  return callCommand('feriados_list', { anio })
}

export function syncFeriados(anios: number[]): Promise<FeriadoSyncResult> {
  return callCommand('feriados_sync', { anios })
}

/** Every write returns the year's calendar, so the caller needs no second request. */
export function addFeriado(dto: FeriadoInput): Promise<Feriado[]> {
  return callCommand('feriados_add', { dto })
}

export function deleteFeriado(fecha: CivilDate): Promise<Feriado[]> {
  return callCommand('feriados_delete', { fecha })
}

File: crates/certaro-application/src/dtos/feriados.rs
//! Contract of the `feriados` module. See `docs/11-contratos-tauri.md` §5.13.

use chrono::NaiveDate;
use certaro_domain::entities::Feriado;
use certaro_domain::OrigenFeriado;
use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct FeriadoDto {
    pub fecha: NaiveDate,
    pub nombre: String,
    pub tipo: Option<String>,
    pub origen: OrigenFeriado,
}

impl From<Feriado> for FeriadoDto {
    fn from(f: Feriado) -> Self {
        Self {
            fecha: f.fecha,
            nombre: f.nombre,
            tipo: f.tipo,
            origen: f.origen,
        }
    }
}

#[derive(Debug, Clone, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct FeriadoInput {
    pub fecha: NaiveDate,
    pub nombre: String,
}

#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct FeriadoSyncResult {
    pub agregados: u64,
    pub total: u64,
    /// Years the provider could not be reached for. The sync is not an error: the calendar simply
    /// stays as it was.
    pub anios_con_error: u32,
}
